sinks/cache/cache_impl.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"sync"
+	"time"
+
+	source_api "github.com/GoogleCloudPlatform/heapster/sources/api"
+)
+
+type containerElement struct {
+	Metadata
+	metrics TimeStore
+}
+
+type podElement struct {
+	Metadata
+	// map of container name to container element.
+	containers map[string]*containerElement
+	// TODO: Cache history of Spec and Status.
+}
+
+type nodeElement struct {
+	node *containerElement
+	// FreeContainers refers to all the containers in a node
+	// that do not belong to a pod.
+	freeContainers map[string]*containerElement
+	// TODO: Cache history of Spec and Status.
+}
+
+type realCache struct {
+	bufferDuration time.Duration
+	// Map of pod UIDs to pod cache entry.
+	pods map[string]*podElement
+	// Map of node hostnames to node cache entry.
+	nodes map[string]*nodeElement
+	lock  sync.RWMutex
+}
+
+const (
+	gcDuration    = time.Second
+	rootContainer = "/"
+	machine       = "machine"
+)
+
+func (rc *realCache) newcontainerElement() *containerElement {
```

Comment:
nit: newContainerElement, same below with the other three. The function is still internal so it probably differentiates itself from a possible NewContainerElement